common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -1565,6 +1567,113 @@ private static boolean contains(CharSequence a, CharSequence b, CharEqualityComp
         return false;
     }
 
+    /**
+     * @param cs the {@code CharSequence} to be processed
+     * @param ignoreCase specifies if case should be ignored.
+     * @param csStart the starting offset in the {@code cs} CharSequence
+     * @param string the {@code CharSequence} to compare.
+     * @param start the starting offset in the specified {@code string}.
+     * @param length the number of characters to compare.
+     * @return {@code true} if the ranges of characters are equal, {@code false} otherwise.
+     */
+    public static boolean regionMatches(final CharSequence cs, final boolean ignoreCase, final int csStart,
+                                        final CharSequence string, final int start, final int length) {
+        if (cs == null || string == null) {
+            return false;
+        }
+
+        if (cs instanceof String && string instanceof String) {
+            return ((String) cs).regionMatches(ignoreCase, csStart, (String) string, start, length);
+        } else if (cs instanceof AsciiString) {
+            return ((AsciiString) cs).regionMatches(ignoreCase, csStart, string, start, length);
+        } else {
+            //general purpose implementation for CharSequences
+            if (csStart < 0 || length > cs.length() - csStart) {
+                return false;
+            }
+            if (start < 0 || length > string.length() - start) {
+                return false;
+            }
+
+            int csIndex = csStart;
+            int csEnd = csIndex + length;
+            int stringIndex = start;
+
+            while (csIndex < csEnd) {
+                char c1 = cs.charAt(csIndex++);
+                char c2 = string.charAt(stringIndex++);
+
+                if (c1 == c2) {
+                    continue;
+                }
+
+                if (ignoreCase) {
+                    //For motivation, why we need two checks, see comment in String#regionMatches
```

Comment:
@Scottmitch Interesting question. For new HttpUtil method - definitely no. But in this case this method will become non universal. I think, let's do it this way:
1) Left this methods as it is.
2) Make two more optimized methods - regionMatchesAscii and indexOfAscii.
3) In regionMatchedAscii will be only
```
Character.toLowerCase(c1) != Character.toLowerCase(c2)
```
and I will delete this
```
if (cs instanceof String && string instanceof String) {
    return ((String) cs).regionMatches(ignoreCase, csStart, (String) string, start, length);
}
```
Because currently we will actually go to String implementation and will make two useless checks.

What do you think?